pkg/bloomgateway/sharding.go

```diff
@@ -0,0 +1,162 @@
+	BloomGatewayShardSize(tenantID string) int
+}
+
+type ShardingStrategy interface {
+	// FilterTenants whose indexes should be loaded by the index gateway.
+	// Returns the list of user IDs that should be synced by the index gateway.
+	FilterTenants(ctx context.Context, tenantIDs []string) ([]string, error)
+	FilterBlocks(ctx context.Context, tenantID string, blockRefs []BlockRef) ([]BlockRef, error)
+}
+
+type ShuffleShardingStrategy struct {
+	r            ring.ReadRing
+	limits       Limits
+	instanceAddr string
+	instanceID   string
+	logger       log.Logger
+}
+
+func NewShuffleShardingStrategy(r ring.ReadRing, l Limits, instanceAddr, instanceID string, logger log.Logger) *ShuffleShardingStrategy {
+	return &ShuffleShardingStrategy{
+		r:            r,
+		limits:       l,
+		instanceAddr: instanceAddr,
+		instanceID:   instanceID,
+		logger:       logger,
+	}
+}
+
+// FilterTenants implements ShardingStrategy.
+func (s *ShuffleShardingStrategy) FilterTenants(_ context.Context, tenantIDs []string) ([]string, error) {
+	// As a protection, ensure the bloom gateway instance is healthy in the ring. It could also be missing
+	// in the ring if it was failing to heartbeat the ring and it got remove from another healthy bloom gateway
+	// instance, because of the auto-forget feature.
+	if set, err := s.r.GetAllHealthy(BlocksOwnerSync); err != nil {
+		return nil, err
+	} else if !set.Includes(s.instanceAddr) {
+		return nil, errGatewayUnhealthy
+	}
+
+	var filteredIDs []string
+
+	for _, tenantID := range tenantIDs {
+		subRing := GetShuffleShardingSubring(s.r, tenantID, s.limits)
+
+		// Include the user only if it belongs to this bloom gateway shard.
+		if subRing.HasInstance(s.instanceID) {
+			filteredIDs = append(filteredIDs, tenantID)
+		}
+	}
+
+	return filteredIDs, nil
+}
+
+func getBucket(rangeMin, rangeMax, pos uint64) int {
+	return 0
+}
+
+// FilterBlocks implements ShardingStrategy.
+func (s *ShuffleShardingStrategy) FilterBlocks(ctx context.Context, tenantID string, blockRefs []BlockRef) ([]BlockRef, error) {
+	filteredBlockRefs := make([]BlockRef, 0, len(blockRefs))
+
+	subRing := GetShuffleShardingSubring(s.r, tenantID, s.limits)
+
+	bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+	var rs ring.ReplicationSet
+	var err error
+
+	for _, blockRef := range blockRefs {
+		rs, err = subRing.Get(uint32(blockRef.FromFp), BlocksOwnerSync, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			return nil, err
+		}
+		// Include the block only if it belongs to this bloom gateway shard.
+		if rs.Includes(s.instanceID) {
+			filteredBlockRefs = append(filteredBlockRefs, blockRef)
+			continue
+		}
+
+		rs, err = subRing.Get(uint32(blockRef.ThroughFp), BlocksOwnerSync, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			return nil, err
+		}
+		// Include the block only if it belongs to this bloom gateway shard.
+		if rs.Includes(s.instanceID) {
+			filteredBlockRefs = append(filteredBlockRefs, blockRef)
+			continue
+		}
```

Comment:
If we include the block by two factors: 1) if `FromFp` belongs to this instance 2) if ThroughFp belongs to this instance, it means that two separate bloomGw might own the same block, one by `FromFp` another one by   `ThroughFp` .
We should check only `FromFp` , it's enough to identify the owner

Reply:
It is actually OK that multiple instances own the same block, if the same block span into the key ranges from both instances.

![screenshot_20231006_082906](https://github.com/grafana/loki/assets/281260/2d5a4c1c-eeed-460e-a525-7ab58198ec1b)